modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/MetricsCompactionTest.java

```diff
@@ -0,0 +1,276 @@
+
+        sendDiscoMessage(srvSpi, createMetricsMessage(crd)); // Dummy message for blocking GridWorker.
+        sendDiscoMessage(srvSpi, msgLap1T1);
+        sendDiscoMessage(srvSpi, createMetricsMessage(crd));
+
+        List<TcpDiscoveryMetricsUpdateMessage> metricMsgT1 = metricsMessages(extractQueue(srvSpi));
+
+        assertEquals(metricMsgT1.size(), 2);
+        assertTrue(metricMsgT1.stream().anyMatch(m -> m.passedLaps(srv.localNode().id()) == 1));
+        assertTrue(metricMsgT1.stream().anyMatch(m -> m.passedLaps(srv.localNode().id()) == 0));
+
+        // Send two duplicates with new version of data.
+        TcpDiscoveryMetricsUpdateMessage msgLap1T2 = createMetricsMessage(crd);
+        msgLap1T2.setMetrics(srv.localNode().id(), new ClusterMetricsSnapshot());
+        msgLap1T2.setMetrics(TEST_UUID, new ClusterMetricsSnapshot()); // Marker
+
+        TcpDiscoveryMetricsUpdateMessage msg0T2 = createMetricsMessage(crd);
+        msg0T2.setMetrics(TEST_UUID, new ClusterMetricsSnapshot());
+
+        sendDiscoMessage(srvSpi, msg0T2);
+        sendDiscoMessage(srvSpi, msgLap1T2);
+
+        List<TcpDiscoveryMetricsUpdateMessage> metricMsgT2 = metricsMessages(extractQueue(srvSpi));
+
+        assertEquals(metricMsgT2.size(), 2);
+        assertTrue(metricMsgT2.stream().anyMatch(m -> m.passedLaps(srv.localNode().id()) == 1));
+        assertTrue(metricMsgT2.stream().anyMatch(m -> m.passedLaps(srv.localNode().id()) == 0));
+
+        record.set(true);
+        latchs.put(0, new CountDownLatch(1));
+        latchs.get(1).countDown();
+
+        assertTrue(GridTestUtils.waitForCondition(() -> records.size() == 2, 10_000));
+        assertTrue(records.stream().anyMatch(m -> m.passedLaps(srv.localNode().id()) == 1 && m.metrics().containsKey(TEST_UUID)));
+        assertTrue(records.stream().anyMatch(m -> m.passedLaps(srv.localNode().id()) == 0 && m.metrics().containsKey(TEST_UUID)));
+    }
+
+    /**
+     *
+     */
+    @Test
+    public void testQueueDoesNotContainMoreTwoMetricMessages() throws Exception {
+        IgniteEx crd = startGrid(0);
+
+        IgniteEx srv = startGrid(1);
+        TcpDiscoverySpi srvSpi = (TcpDiscoverySpi)srv.configuration().getDiscoverySpi();
+
+        AtomicBoolean record = new AtomicBoolean(false);
+        List<TcpDiscoveryMetricsUpdateMessage> records = Collections.synchronizedList(new ArrayList<>());
+
+        srvSpi.addSendMessageListener(msg -> {
+            if(msg instanceof TcpDiscoveryMetricsUpdateMessage) {
+                try {
+                    if(record.get())
+                        records.add((TcpDiscoveryMetricsUpdateMessage)msg);
+
+                    latchs.get(1).await();
+                }
+                catch (InterruptedException ignore) {
+                }
+            }
+        });
+
+        latchs.put(1, new CountDownLatch(1)); // Blocks srv discovery for sending
+
+        Thread.sleep(1_000);
```

Comment:
Why do we need to sleep here for 1 sec?

Reply:
I added another type of dummy message, it should allow remove it.